File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItemBuilder.java
package org.schabi.newpipe.player.playqueue;

import android.content.Context;
import android.text.TextUtils;
import android.view.MotionEvent;
import android.view.View;

import org.schabi.newpipe.util.Localization;
import org.schabi.newpipe.util.image.PicassoHelper;
import org.schabi.newpipe.util.ServiceHelper;

public class PlayQueueItemBuilder {
    private static final String TAG = PlayQueueItemBuilder.class.toString();
    private OnSelectedListener onItemClickListener;

    public PlayQueueItemBuilder(final Context context) {
    }

    public void setOnSelectedListener(final OnSelectedListener listener) {
        this.onItemClickListener = listener;
    }

    public void buildStreamInfoItem(final PlayQueueItemHolder holder, final PlayQueueItem item) {
        if (!TextUtils.isEmpty(item.getTitle())) {
            holder.itemVideoTitleView.setText(item.getTitle());
        }
        holder.itemAdditionalDetailsView.setText(Localization.concatenateStrings(item.getUploader(),
                ServiceHelper.getNameOfServiceById(item.getServiceId())));

        if (item.getDuration() > 0) {
            holder.itemDurationView.setText(Localization.getDurationString(item.getDuration()));
        } else {
            holder.itemDurationView.setVisibility(View.GONE);
        }

        PicassoHelper.loadThumbnail(item.getThumbnails()).into(holder.itemThumbnailView);

        holder.itemRoot.setOnClickListener(view -> {
            if (onItemClickListener != null) {
                onItemClickListener.selected(item, view);
            }
        });

        holder.itemRoot.setOnLongClickListener(view -> {
            if (onItemClickListener != null) {
                onItemClickListener.held(item, view);
                return true;
            }
            return false;
        });

        holder.itemHandle.setOnTouchListener(getOnTouchListener(holder));
    }

    private View.OnTouchListener getOnTouchListener(final PlayQueueItemHolder holder) {
        return (view, motionEvent) -> {
            view.performClick();
            if (motionEvent.getActionMasked() == MotionEvent.ACTION_DOWN
                    && onItemClickListener != null) {
                onItemClickListener.onStartDrag(holder);
            }
            return false;
        };
    }

    public interface OnSelectedListener {
        void selected(PlayQueueItem item, View view);

        void held(PlayQueueItem item, View view);

        void onStartDrag(PlayQueueItemHolder viewHolder);
    }
}

File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItemBuilder.kt
package org.schabi.newpipe.player.playqueue

import android.content.Context
import android.text.TextUtils
import android.view.MotionEvent
import android.view.View
import android.view.View.OnLongClickListener
import android.view.View.OnTouchListener
import org.schabi.newpipe.util.Localization
import org.schabi.newpipe.util.ServiceHelper
import org.schabi.newpipe.util.image.PicassoHelper

class PlayQueueItemBuilder(context: Context?) {
    private var onItemClickListener: OnSelectedListener? = null
    fun setOnSelectedListener(listener: OnSelectedListener?) {
        onItemClickListener = listener
    }

    fun buildStreamInfoItem(holder: PlayQueueItemHolder, item: PlayQueueItem?) {
        if (!TextUtils.isEmpty(item.getTitle())) {
            holder.itemVideoTitleView.setText(item.getTitle())
        }
        holder.itemAdditionalDetailsView.setText(Localization.concatenateStrings(item.getUploader(),
                ServiceHelper.getNameOfServiceById(item.getServiceId())))
        if (item.getDuration() > 0) {
            holder.itemDurationView.setText(Localization.getDurationString(item.getDuration()))
        } else {
            holder.itemDurationView.setVisibility(View.GONE)
        }
        PicassoHelper.loadThumbnail(item.getThumbnails()).into(holder.itemThumbnailView)
        holder.itemRoot.setOnClickListener(View.OnClickListener({ view: View? ->
            if (onItemClickListener != null) {
                onItemClickListener!!.selected(item, view)
            }
        }))
        holder.itemRoot.setOnLongClickListener(OnLongClickListener({ view: View? ->
            if (onItemClickListener != null) {
                onItemClickListener!!.held((item)!!, view)
                return@setOnLongClickListener true
            }
            false
        }))
        holder.itemHandle.setOnTouchListener(getOnTouchListener(holder))
    }

    private fun getOnTouchListener(holder: PlayQueueItemHolder): OnTouchListener {
        return OnTouchListener({ view: View, motionEvent: MotionEvent ->
            view.performClick()
            if ((motionEvent.getActionMasked() == MotionEvent.ACTION_DOWN
                            && onItemClickListener != null)) {
                onItemClickListener!!.onStartDrag(holder)
            }
            false
        })
    }

    open interface OnSelectedListener {
        fun selected(item: PlayQueueItem?, view: View?)
        fun held(item: PlayQueueItem, view: View?)
        fun onStartDrag(viewHolder: PlayQueueItemHolder?)
    }

    companion object {
        private val TAG: String = PlayQueueItemBuilder::class.java.toString()
    }
}
